test/test_util.js

```diff
@@ -1,213 +1,109 @@
-      newCaretPosition = selectionStart;
-      newValue =
-        selectionStart !== selectionEnd
-          ? currentValue.substring(0, selectionStart) +
-            currentValue.substring(selectionEnd, currentValue.length)
-          : currentValue.substring(0, selectionStart) +
-            currentValue.substring(selectionStart + 1, currentValue.length);
-    } else {
-      newCaretPosition = selectionStart + key.length;
-      newValue =
-        selectionStart !== selectionEnd
-          ? currentValue.substring(0, selectionStart) +
-            key +
-            currentValue.substring(selectionEnd, currentValue.length)
-          : currentValue.substring(0, selectionStart) +
-            key +
-            currentValue.substring(selectionStart, currentValue.length);
-    }
+  const specialKeys = ['{Backspace}', '{Delete}', '{ArrowLeft}', '{ArrowRight}'];
 
-    const changeEvent = getEvent(
-      {
-        persist: persist.bind(null, 'change'),
-        key,
-      },
-      {
-        value: newValue,
-        selectionStart: newCaretPosition,
-        selectionEnd: newCaretPosition,
-        setSelectionRange: setSelectionRange || noop,
-        focus: noop,
-      },
-    );
-    input.simulate('change', changeEvent);
-  }
-}
+  if (specialKeys.includes(key)) {
+    if (start === end) {
+      await input.focus();
+      await input.setSelectionRange(start, end);
 
-export async function render(elm) {
-  const view = testRender(elm);
-  const input = await view.getByRole('textbox');
-  return { ...view, view: view, input };
-}
+      await user.keyboard(key);
+    } else {
+      let newValue;
 
-export function simulateNativeKeyInput(input, key, selectionStart = 0, selectionEnd = 0) {
-  input.setSelectionRange(selectionStart, selectionEnd);
-  userEvent.type(input, key);
-}
+      if (key === '{Backspace}') {
+        newValue = v.slice(0, start) + v.slice(end, v.length);
+      } else if (key === '{Delete}') {
+        newValue = v.slice(0, start) + v.slice(end, v.length);
+      }
 
-export function simulatePaste(input, test, selectionStart = 0, selectionEnd = 0) {
-  input.setSelectionRange(selectionStart, selectionEnd);
-  userEvent.paste(input, test);
-}
+      fireEvent.change(input, { target: { value: newValue } });
+    }
+  } else {
+    if (start === end) {
+      if (options?.eventType === 'keyboard') {
+        input.focus();
+        input.selectionStart = start;
+        input.selectionEnd = end;
+
+        await user.keyboard(key);
+      } else {
+        await user.type(input, key, { initialSelectionStart: start, initialSelectionEnd: end });
+      }
+    } else {
+      let newValue;
 
-export function simulateNativeMouseUpEvent(input, selectionStart) {
-  input.setSelectionRange(selectionStart, selectionStart);
-  userEvent.click(input);
+      newValue = v.slice(0, start) + key + v.slice(end, v.length);
+      fireEvent.change(input, { target: { value: newValue } });
+      end = start;
+    }
```

Comment:
Q: Why do need custom handling here?
Won't just this work.
```
user.type(input, key, { initialSelectionStart: start, initialSelectionEnd: end });
```

Or
```
input.setSelectionRange(selectionStart, selectionEnd);
userEvent.type(input, key);
```

Reply:
Actually, I tried `await user.type(input, key, { initialSelectionStart: start, initialSelectionEnd: end });`, looks like in some cases it's not working, basically due to JSDOM, the selected text is not getting cleared first. Need to dig more, but for now we can keep what you have added.

Reply:
Haven't pushed the relevant commits yet but I have already fixed this locally.